rfcs/default_idp_configurable/default_idp_configurable.md

```diff
@@ -0,0 +1,393 @@
+
+# Design Summary
+The general approach proposed by this RFC is to add some configurations to allow Service Workbench to integrate with Keycloak. This will make Service Workbench available in the regions where Cognito User Pool is not available.
+
+## Detail Design
+The design will be divided into two parts: Deployment Process and Runtime Process.
+
+### Deployment Process design
+We keep the current highlevel design that invoking lambda function(postDeployment) to create the resources related with the authentication feature. But which resources are created is determined by configuration as below in "main/config/settings/${STAGE_NAME}.yml" file:
+```
+defaultIDPType: keycloak/cognito
+```
+If user choose "keycloak", because keycloak resources are not created easily during the Service Workbench deployment process, there are some prerequisites that user need to do:
+1. User should have a deployed keycloak.
+2. Create a realm for Service Workbench or use existed realm.
+3. Create a client for Service Workbench.
+4. Create root user in Keycloak.
+
+We will add following configurations options for keycloak integration.
+```
+keyCloakRealm: xxx
+keyCloakClientId: xxx
+keyCloakUrl: xxx
+rootUserEmail: xxx
+rootUserFirstName: xxx
+rootUserLastName: xxx
+```
+![deployment_keycloak](./swb_auth_deploy_keycloak.drawio.png)
+
+#### AddAuthProviders
+In AddAuthProviders.execute(), run keycloak or cognito logic through defaultIDPType config. Code as below:
+```
+  async execute() {
+    const defaultIDPType = this.settings.get(settingKeys.defaultIDPType);
+    switch (defaultIDPType) {
+      case 'Keycloak':
+        await this.addKeycloakAuthenticationProviderWithSamlFederation();
+        break;
+      case 'Cognito':
+        await this.addCognitoAuthenticationProviderWithSamlFederation();
+        break;
+      default:
+        console.log(`Sorry, we are out of ${defaultIDPType}.`);
+    } 
+  }
+```
+
+In the "addKeycloakAuthenticationProviderWithSamlFederation()" method, it need to add generate keycloak **Auth Provider Config** and save it into DDB.
+The keycloak **Auth Provider Config** should be similar to the CognitoAuthProviderConfig, also contains the following three parts:
+  - Keycloak information:
+```
+{
+  ...
+  "clientId": 
+  "signInUri": 
+  "signOutUri": 
+  "authCodeTokenExchangeUri":
+  ...
+}
+```
+
+- Federated information:
+```
+{
+  ...
+  "federatedIdentityProviders": [
+    {
+      "id":
+      "name": 
+      "displayName":
+    }
+  ],
+  ...
+}
+```
+
+- Implementation configurations:
+```
+{
+  ...
+  "impl": {
+    ...
+    "tokenValidatorLocator": "locator:service:keycloakAuthenticationProviderService/validateToken",
+    "tokenRevokerLocator": "locator:service:keycloakAuthenticationProviderService/revokeToken"
+    ...
+  }
+  ...
+}
+```
+
+#### KeycloakAuthenticationProviderService
```

Comment:
Is it a runtime service for authentication?

Reply:
Yes, it is runtime service for authenctication, I will move it to runtime part.